src/log.js

```diff
@@ -1,153 +1,389 @@
+   * @private
+   * @param {Entry} entry Entry to be inserted
+   * @returns {Entry}
+   */
+  static _insert(ipfs, log, entry) {
+    const hashes = log.items.map((f) => f.hash)
+    // If entry is already in the log, don't insert
+    if (hashes.includes(entry.hash)) return entry
+    // Find the item's parents' indices
+    const indices = entry.next.map((next) => hashes.indexOf(next))
+    // Find the largest index (latest parent)
+    const index = indices.length > 0 ? Math.max(Math.max.apply(null, indices) + 1, 0) : 0
+    // Insert
+    log.items.splice(index, 0, entry)
+    return entry
+  }
+
+  /**
+   * Fetch log entries recursively
+   * @private
+   * @param {IPFS} [ipfs] An IPFS instance
+   * @param {string} [hash] Multihash of the entry to fetch
+   * @param {string} [parent] Parent of the node to be fetched
+   * @param {Object} [all] Entries to skip
+   * @param {Number} [amount=-1] How many entries to fetch.
+   * @param {Number} [depth=0] Current depth of the recursion
+   * @param {function(hash, entry, parent, depth)} onProgressCallback
+   * @returns {Promise<Array<Entry>>}
+   */
+  static _fetchRecursive(ipfs, hash, all = {}, amount = -1, depth = 0, parent = null, onProgressCallback = () => {}) {
+    // If the given hash is already fetched
+    // or if we're at maximum depth, return
+    if (all[hash] || (depth >= amount && amount > 0)) {
+      return Promise.resolve([])
+    }
+    // Create the entry and add it to the result
+    return Entry.fromMultihash(ipfs, hash)
+      .then((entry) => {
+        all[hash] = entry
+        onProgressCallback(hash, entry, parent, depth)
+        const fetch = (hash) => LogUtils._fetchRecursive(ipfs, hash, all, amount, depth + 1, entry, onProgressCallback)
+        return mapSeries(entry.next, fetch, { concurrency: 1 })
+          .then((res) => res.concat([entry]))
+          .then((res) => res.reduce((a, b) => a.concat(b), [])) // flatten the array
       })
-      .then((items) => Object.assign(options, { items: items }))
-      .then((items) => new Log(ipfs, logData.id, options))
   }
 
-  static findHeads(log) {
-    return log.items
+  /**
+   * Find heads of a log
+   * @private
+   * @param {Log} log
+   * @returns {Array<Entry>}
+   */
+  static _findHeads(log) {
+    return log.items.slice()
       .reverse()
-      .filter((f) => !Log.isReferencedInChain(log, f))
+      .filter((f) => !LogUtils._isReferencedInChain(log, f))
       .map((f) => f.hash)
+      .sort(LogUtils._compare)
   }
 
-  static isReferencedInChain(log, item) {
-    return log.items.reverse().find((e) => Entry.hasChild(e, item)) !== undefined
+  /**
+   * Check if an entry is referenced by another entry in the log
+   * @private
+   * @param {log} [log] Log to search an entry from
+   * @param {Entry} [entry] Entry to search for
+   * @returns {boolean}
+   */
+  static _isReferencedInChain(log, entry) {
+    return log.items.slice().reverse().find((e) => Entry.hasChild(e, entry)) !== undefined
```

Comment:
Interesting, would be curious if that actually helps in perf, because the copy + reverse is not cheap for large arrays I expect. 

In either case might be good to leave a comment, so the next person changing the code knows about this.